File: app/src/main/java/com/jon/cotgenerator/ui/SettingsFragment.java
package com.jon.cotgenerator.ui;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.text.InputType;
import android.widget.EditText;

import androidx.annotation.Nullable;
import androidx.preference.EditTextPreference;
import androidx.preference.ListPreference;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;
import androidx.preference.PreferenceManager;
import androidx.preference.SeekBarPreference;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.jon.cotgenerator.R;
import com.jon.cotgenerator.enums.Protocol;
import com.jon.cotgenerator.utils.InputValidator;
import com.jon.cotgenerator.utils.Key;
import com.jon.cotgenerator.utils.Notify;
import com.jon.cotgenerator.utils.OutputPreset;
import com.jon.cotgenerator.utils.PrefUtils;
import com.jon.cotgenerator.utils.PresetSqlHelper;

import org.apache.commons.collections4.ListUtils;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public class SettingsFragment
        extends PreferenceFragmentCompat
        implements SharedPreferences.OnSharedPreferenceChangeListener,
                   Preference.OnPreferenceChangeListener {
    private SharedPreferences prefs;
    private PresetSqlHelper sqlHelper;

    static SettingsFragment newInstance() {
        return new SettingsFragment();
    }

    private static final String[] PHONE_INPUT = new String[]{
            Key.CENTRE_LATITUDE,
            Key.CENTRE_LONGITUDE,
            Key.ICON_COUNT,
            Key.MOVEMENT_SPEED,
            Key.RADIAL_DISTRIBUTION,
    };

    private static final Map<String, String> SUFFIXES = new HashMap<String, String>() {{
        put(Key.CENTRE_LATITUDE, "degrees");
        put(Key.CENTRE_LONGITUDE, "degrees");
        put(Key.MOVEMENT_SPEED, "mph");
        put(Key.RADIAL_DISTRIBUTION, "metres");
    }};

    private static final Map<String, String> PREFS_REQUIRING_VALIDATION = new HashMap<String, String>() {{
        put(Key.CENTRE_LATITUDE, "Should be a number between -180 and +180");
        put(Key.CENTRE_LONGITUDE, "Should be a number between -90 and +90");
        put(Key.ICON_COUNT, "Should be an integer from 1 to 9999");
        put(Key.MOVEMENT_SPEED, "Should be a positive number");
        put(Key.RADIAL_DISTRIBUTION, "Should be a positive integer");
    }};

    private static final String[] SEEKBARS = new String[]{
            Key.STALE_TIMER,
            Key.TRANSMISSION_PERIOD
    };

    @Override
    public void onCreatePreferences(Bundle savedState, String rootKey) {
        setPreferencesFromResource(R.xml.settings, rootKey);

        EditTextPreference.OnBindEditTextListener phoneInputType = (EditText text) -> text.setInputType(InputType.TYPE_CLASS_PHONE);
        for (final String key : PHONE_INPUT) {
            EditTextPreference pref = findPreference(key);
            if (pref != null) pref.setOnBindEditTextListener(phoneInputType);
        }

        for (final String key : PREFS_REQUIRING_VALIDATION.keySet()) {
            Preference pref = findPreference(key);
            if (pref != null) pref.setOnPreferenceChangeListener(this);
        }
        for (final String key : SEEKBARS) {
            SeekBarPreference seekbar = findPreference(key);
            seekbar.setMin(1); /* I can't set the minimum in the XML for whatever reason, so here it is */
        }
        setPresetPreferenceListeners();
        sqlHelper = new PresetSqlHelper(requireContext());
    }

    @Override
    public void onActivityCreated(@Nullable Bundle state) {
        super.onActivityCreated(state);
        prefs = PreferenceManager.getDefaultSharedPreferences(requireActivity());
        prefs.registerOnSharedPreferenceChangeListener(this);
    }

    @Override
    public void onStart() {
        super.onStart();
        updatePreferences();
    }

    @Override
    public void onResume() {
        super.onResume();
        for (Map.Entry<String, String> entry : SUFFIXES.entrySet()) {
            setPreferenceSuffix(prefs, entry.getKey(), entry.getValue());
        }
        updatePreferences();
    }

    private void updatePreferences() {
        toggleProtocolSettingVisibility();
        toggleDataFormatSettingVisibility();
        setColourPickerActive();
        setPositionPrefsActive();
        updatePresetEntries(Protocol.UDP, Key.UDP_PRESETS);
        updatePresetEntries(Protocol.TCP, Key.TCP_PRESETS);
        Protocol newProtocol = Protocol.fromPrefs(prefs);
        insertPresetAddressAndPort(newProtocol == Protocol.TCP ? Key.TCP_PRESETS : Key.UDP_PRESETS);
    }

    @Override
    public void onDestroy() {
        prefs.unregisterOnSharedPreferenceChangeListener(this);
        sqlHelper.close();
        super.onDestroy();
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences prefs, String key) {
        if (SUFFIXES.containsKey(key)) {
            setPreferenceSuffix(this.prefs, key, SUFFIXES.get(key));
        }
        switch (key) {
            case Key.TRANSMISSION_PROTOCOL:
                toggleProtocolSettingVisibility();
                toggleDataFormatSettingVisibility();
                Protocol newProtocol = Protocol.fromPrefs(prefs);
                insertPresetAddressAndPort(newProtocol == Protocol.TCP ? Key.TCP_PRESETS : Key.UDP_PRESETS);
                break;
            case Key.FOLLOW_GPS_LOCATION:
                setPositionPrefsActive();
                break;
            case Key.TCP_PRESETS:
            case Key.UDP_PRESETS:
                insertPresetAddressAndPort(key);
                break;
            case Key.RANDOM_COLOUR:
                setColourPickerActive();
                break;
            case Key.NEW_PRESET_ADDED:
                updatePresetEntries(Protocol.UDP, Key.UDP_PRESETS);
                updatePresetEntries(Protocol.TCP, Key.TCP_PRESETS);
                break;
        }
    }

    private void setPositionPrefsActive() {
        boolean followGps = PrefUtils.getBoolean(prefs, Key.FOLLOW_GPS_LOCATION);
        findPreference(Key.CENTRE_LATITUDE).setVisible(!followGps);
        findPreference(Key.CENTRE_LONGITUDE).setVisible(!followGps);
    }

    private void insertPresetAddressAndPort(String key) {
        EditTextPreference addressPref = findPreference(Key.DEST_ADDRESS);
        EditTextPreference portPref = findPreference(Key.DEST_PORT);
        ListPreference presetPref = findPreference(key);
        if (addressPref != null && portPref != null && presetPref != null) {
            OutputPreset preset = OutputPreset.fromString(presetPref.getValue());
            if (preset != null) {
                addressPref.setText(preset.address);
                portPref.setText(Integer.toString(preset.port));
            } else {
                presetPref.setValue(null);
                addressPref.setText(null);
                portPref.setText(null);
            }
        }
    }

    private void toggleProtocolSettingVisibility() {
        boolean showUdpSettings = Protocol.fromPrefs(prefs) == Protocol.UDP;
        findPreference(Key.UDP_PRESETS).setVisible(showUdpSettings);
        findPreference(Key.TCP_PRESETS).setVisible(!showUdpSettings);

    private void toggleDataFormatSettingVisibility() {
        /* Data format is only relevant for UDP, since TAK Server only takes XML data */
        boolean showDataFormatSettings = Protocol.fromPrefs(prefs) == Protocol.UDP;
        Preference dataFormat = findPreference(Key.DATA_FORMAT);
        if (dataFormat != null) {
            dataFormat.setVisible(showDataFormatSettings);
        }
    }

    private void setColourPickerActive() {
        boolean useRandomColours = PrefUtils.getBoolean(prefs, Key.RANDOM_COLOUR);
        Preference colourPicker = findPreference(Key.TEAM_COLOUR);
        if (colourPicker != null) {
            colourPicker.setVisible(!useRandomColours);
        }
    }

    @Override
    public boolean onPreferenceChange(Preference pref, Object newValue) {
        final String str = (String) newValue;
        boolean result = true;
        switch (pref.getKey()) {
            case Key.CENTRE_LATITUDE:
                result = InputValidator.validateDouble(str, -90.0, 90.0);
                break;
            case Key.CENTRE_LONGITUDE:
                result = InputValidator.validateDouble(str, -180.0, 180.0);
                break;
            case Key.ICON_COUNT:
                result = InputValidator.validateInt(str, 1, 9999);
                break;
            case Key.RADIAL_DISTRIBUTION:
            case Key.TRANSMISSION_PERIOD:
                result = InputValidator.validateInt(str, 1, null);
                break;
            case Key.MOVEMENT_SPEED:
                result = InputValidator.validateDouble(str, 0.0, null);
                break;
        }
        if (!result) {
            Notify.red(requireView(), "Invalid input: " + str + ". " + PREFS_REQUIRING_VALIDATION.get(pref.getKey()));
        }
        return result;
    }

    private void setPreferenceSuffix(final SharedPreferences prefs, final String key, final String suffix) {
        Preference pref = findPreference(key);
        if (pref != null) {
            String val = prefs.getString(key, "");
            pref.setSummary(String.format(Locale.ENGLISH, "%s %s", val, suffix));
        }
    }

    private void setPresetPreferenceListeners() {
        Preference addPreference = findPreference(Key.ADD_NEW_PRESET);
        if (addPreference != null) {
            addPreference.setOnPreferenceClickListener(clickedPref -> {
                NewPresetDialogCreator.show(requireContext(), requireView(), prefs, sqlHelper);
                return true;
            });
        }
        Preference deletePreference = findPreference(Key.DELETE_PRESETS);
        if (deletePreference != null) {
            deletePreference.setOnPreferenceClickListener(clickedPref -> {
                deletePresetDialog();
                return true;
            });
        }
    }

    private void updatePresetEntries(Protocol protocol, String key) {
        List<OutputPreset> defaults = (protocol == Protocol.TCP) ? OutputPreset.tcpDefaults() : OutputPreset.udpDefaults();
        List<OutputPreset> presets = ListUtils.union(defaults, sqlHelper.getAllPresets(protocol));
        List<String> entries = OutputPreset.getAliases(presets);
        List<String> entryValues = new ArrayList<>();
        for (OutputPreset preset : presets) {
            entryValues.add(preset.toString());
        }
        ListPreference preference = findPreference(key);
        if (preference != null) {
            String previousValue = preference.getValue();
            preference.setEntries(Arrays.copyOf(entries.toArray(), entries.toArray().length, String[].class));
            preference.setEntryValues(Arrays.copyOf(entryValues.toArray(), entryValues.toArray().length, String[].class));
            preference.setValue(previousValue);
        }
    }

    private void deletePresetDialog() {
        new MaterialAlertDialogBuilder(requireContext())
                .setTitle("Delete Presets")
                .setMessage("Clear all custom output presets? The built-in defaults will still remain.")
                .setPositiveButton(android.R.string.ok, (dialog, buttonId) -> {
                    resetPresetPreference(Key.UDP_PRESETS, OutputPreset.udpDefaults());
                    resetPresetPreference(Key.TCP_PRESETS, OutputPreset.tcpDefaults());
                    if (PresetSqlHelper.deleteDatabase()) {
                        Notify.green(requireView(), "Successfully deleted presets");
                    } else {
                        Notify.red(requireView(), "Failed to delete presets");
                    }
                }).setNegativeButton(android.R.string.cancel, (dialog, buttonId) -> dialog.dismiss())
                .show();
    }

    private void resetPresetPreference(String prefKey, List<OutputPreset> defaults) {
        List<String> entries = OutputPreset.getAliases(defaults);
        List<String> values = new ArrayList<>();
        for (OutputPreset preset : defaults) {
            values.add(preset.toString());
        }
        ListPreference preference = findPreference(prefKey);
        if (preference != null) {
            preference.setEntries(Arrays.copyOf(entries.toArray(), entries.size(), String[].class));
            preference.setEntryValues(Arrays.copyOf(values.toArray(), values.size(), String[].class));
            preference.setValueIndex(0);
        }
    }
}

File: common/src/main/java/com/jon/common/ui/main/SettingsFragment.kt
package com.jon.common.ui.main

import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.os.Bundle
import android.text.InputType
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.navigation.fragment.findNavController
import androidx.preference.*
import androidx.preference.EditTextPreference.OnBindEditTextListener
import com.jon.common.di.IUiResources
import com.jon.common.prefs.CommonPrefs
import com.jon.common.prefs.PrefPair
import com.jon.common.prefs.RefreshCallsignPreference
import com.jon.common.presets.OutputPreset
import com.jon.common.repositories.IPresetRepository
import com.jon.common.utils.InputValidator
import com.jon.common.utils.Notify
import com.jon.common.utils.Protocol
import com.jon.common.utils.safelyNavigate
import timber.log.Timber
import javax.inject.Inject
import kotlin.collections.ArrayList

abstract class SettingsFragment : PreferenceFragmentCompat(),
        OnSharedPreferenceChangeListener,
        Preference.OnPreferenceChangeListener {

    private val navController by lazy { findNavController() }

    @Inject
    lateinit var presetRepository: IPresetRepository

    @Inject
    protected lateinit var prefs: SharedPreferences

    @Inject
    protected lateinit var uiResources: IUiResources

    protected open fun getPhoneInputKeys() = mutableListOf<String>(
            /* blank, all phone inputs are in Generator only */
    )

    protected open fun getSuffixes() = mutableMapOf<String, String>(
            /* blank, all suffixes are in Generator only */
    )

    protected open fun getPrefValidationRationales() = mutableMapOf(
            CommonPrefs.CALLSIGN.key to "Contains invalid character(s)"
    )

    protected open fun getSeekbarKeys() = mutableListOf(
            CommonPrefs.STALE_TIMER.key,
            CommonPrefs.TRANSMISSION_PERIOD.key
    )

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        Timber.d("onCreateView")
        setHasOptionsMenu(false)
        return super.onCreateView(inflater, container, savedInstanceState)
    }

    override fun onCreatePreferences(savedState: Bundle?, rootKey: String?) {
        Timber.d("onCreatePreferences")
        setPreferencesFromResource(uiResources.settingsXmlId, rootKey)

        val callsignPreference = findPreference<RefreshCallsignPreference>(CommonPrefs.CALLSIGN.key)
        callsignPreference?.setUiResources(uiResources)

        /* Set numeric input on numeric fields */
        val phoneInputType = OnBindEditTextListener { it.inputType = InputType.TYPE_CLASS_PHONE }
        getPhoneInputKeys().forEach {
            findPreference<EditTextPreference>(it)?.setOnBindEditTextListener(phoneInputType)
        }

        /* Register to intercept preference changes */
        getPrefValidationRationales().keys.forEach {
            findPreference<Preference>(it)?.onPreferenceChangeListener = this
        }

        /* Set the minimum seekbar values to 1, since the XML attribute for this doesn't work */
        getSeekbarKeys().forEach {
            findPreference<SeekBarPreference>(it)?.min = 1
        }

        /* Launch a new activity when clicking "Edit Presets" */
        findPreference<Preference>(CommonPrefs.EDIT_PRESETS)?.let {
            it.onPreferenceClickListener = Preference.OnPreferenceClickListener {
                Timber.d("Clicked custom presets")
                navController.safelyNavigate(uiResources.mainToListDirections)
                true
            }
        }
    }

    override fun onActivityCreated(state: Bundle?) {
        Timber.d("onActivityCreated")
        super.onActivityCreated(state)
        prefs.registerOnSharedPreferenceChangeListener(this)
    }

    override fun onStart() {
        Timber.d("onStart")
        super.onStart()
        updatePreferences()
    }

    override fun onResume() {
        Timber.d("onResume")
        super.onResume()
        for ((key, value) in getSuffixes()) {
            setPreferenceSuffix(prefs, key, value)
        }
    }

    protected open fun updatePreferences() {
        Timber.d("updatePreferences")
        toggleProtocolSettingVisibility()
        toggleDataFormatSettingVisibility()
        updatePresetPreferences()
        insertPresetAddressAndPort()
    }

    override fun onDestroy() {
        Timber.d("onDestroy")
        prefs.unregisterOnSharedPreferenceChangeListener(this)
        super.onDestroy()
    }

    override fun onSharedPreferenceChanged(prefs: SharedPreferences, key: String) {
        Timber.d("onSharedPreferenceChanged %s", key)
        val suffixes = getSuffixes()
        if (suffixes.containsKey(key)) {
            setPreferenceSuffix(this.prefs, key, suffixes[key])
        }
        when (key) {
            CommonPrefs.TRANSMISSION_PROTOCOL.key -> {
                toggleProtocolSettingVisibility()
                toggleDataFormatSettingVisibility()
                insertPresetAddressAndPort()
            }
            CommonPrefs.SSL_PRESETS.key, CommonPrefs.TCP_PRESETS.key, CommonPrefs.UDP_PRESETS.key ->
                insertPresetAddressAndPort()
        }
    }

    protected fun <T> setPrefVisibleIfCondition(pref: PrefPair<T>, condition: Boolean) {
        Timber.d("setPrefVisibleIfCondition %s %s", pref.key, condition)
        findPreference<Preference>(pref.key)?.isVisible = condition
    }

    private fun toggleProtocolSettingVisibility() {
        val protocol = Protocol.fromPrefs(prefs)
        Timber.d("toggleProtocolSettingVisibility %s", protocol.name)
        setPrefVisibleIfCondition(CommonPrefs.SSL_PRESETS, protocol == Protocol.SSL)
        setPrefVisibleIfCondition(CommonPrefs.TCP_PRESETS, protocol == Protocol.TCP)
        setPrefVisibleIfCondition(CommonPrefs.UDP_PRESETS, protocol == Protocol.UDP)
    }

    private fun toggleDataFormatSettingVisibility() {
        /* Data format is only relevant for UDP, since TAK Server only takes XML data */
        val showDataFormatSetting = Protocol.fromPrefs(prefs) == Protocol.UDP
        Timber.d("toggleDataFormatSettingVisibility %s", showDataFormatSetting)
        setPrefVisibleIfCondition(CommonPrefs.DATA_FORMAT, showDataFormatSetting)
    }

    override fun onPreferenceChange(pref: Preference, newValue: Any): Boolean {
        Timber.d("onPreferenceChange %s", newValue.toString())
        val input = newValue as String
        val inputValidator = InputValidator()
        return when (val key = pref.key) {
            CommonPrefs.CALLSIGN.key ->
                errorIfInvalid(input, key, inputValidator.validateCallsign(input))
            CommonPrefs.TRANSMISSION_PERIOD.key ->
                errorIfInvalid(input, key, inputValidator.validateInt(input, 1, null))
            else -> true
        }
    }

    protected fun errorIfInvalid(input: String, key: String?, result: Boolean): Boolean {
        Timber.d("errorIfInvalid %s %s %s", input, key, result)
        if (!result) {
            Notify.red(requireView(), "Invalid input: " + input + ". " + getPrefValidationRationales()[key])
        }
        return result
    }

    private fun setPreferenceSuffix(prefs: SharedPreferences, key: String, suffix: String?) {
        Timber.d("setPreferenceSuffix %s %s", key, suffix)
        findPreference<Preference>(key)?.let {
            val value = prefs.getString(key, "")
            it.summary = "$value $suffix"
        }
    }

    private fun insertPresetAddressAndPort() {
        Timber.d("insertPresetAddressAndPort")
        val addressPref = findPreference<EditTextPreference>(CommonPrefs.DEST_ADDRESS)
        val portPref = findPreference<EditTextPreference>(CommonPrefs.DEST_PORT)
        val presetPref = findPreference<ListPreference>(
                Protocol.fromPrefs(prefs).presetPref.key
        )
        if (addressPref != null && portPref != null && presetPref != null) {
            val preset = OutputPreset.fromString(presetPref.value)
            if (preset != null) {
                addressPref.text = preset.address
                portPref.text = preset.port.toString()
            } else {
                presetPref.value = null
                addressPref.text = null
                portPref.text = null
            }
        }
    }

    private fun updatePresetPreferences() {
        Timber.d("updatePresetPreferences")
        mapOf(
                Protocol.SSL to CommonPrefs.SSL_PRESETS.key,
                Protocol.TCP to CommonPrefs.TCP_PRESETS.key,
                Protocol.UDP to CommonPrefs.UDP_PRESETS.key,
        ).forEach { (protocol, key) ->
            presetRepository.getCustomByProtocol(protocol).observe(viewLifecycleOwner) { customPresets ->
                val allPresets = ArrayList<OutputPreset>().apply {
                    addAll(presetRepository.defaultsByProtocol(protocol))
                    addAll(customPresets)
                }
                updatePresetEntries(allPresets, key)
            }
        }
    }

    private fun updatePresetEntries(presets: List<OutputPreset>, key: String) {
        Timber.d("updatePresetEntries %s", key)
        val entries = presets.map { it.alias }.toTypedArray()
        val entryValues = presets.map { it.toString() }.toTypedArray()
        findPreference<ListPreference>(key)?.let {
            val previousValue = it.value
            it.entries = entries
            it.entryValues = entryValues
            if (entryValues.contains(previousValue)) {
                it.value = previousValue
            } else {
                it.setValueIndex(0)
            }
        }
    }

}